### PR TITLE
Phpipam upgrade

### PR DESCRIPTION
**Pull request by dugdigg**
The routines that were used by autoip.rb plugin have changed in version 1.21 of phpIPAM.  I have reworked autoip.rb to make use of an existing phpIPAM api controller.

See also https://github.com/phpipam/phpipam/pull/509

### CODE REVIEW

File: plugins/autoip.rb [hunk 1]
```diff
@@ -35,29 +36,104 @@ def self.pre_validate options
...
+
+      # Authenticate and get a token for further operations
+
+      puts "Requesting auth token for #{username}"
+      uri = options[:add_uri].gsub(/APIAPP/, {
+        'APIAPP' => options[:apiapp],
+      })
+      uri = uri 
+      uri = URI.escape(uri)
+      uri = URI.parse(uri.concat("user/"))
```
Comment (dugdigg) [inline plugins/autoip.rb]: when reversing these uri assignments I would get the error below.

/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/uri/common.rb:304:in `escape': undefined method `gsub' for #<URI::HTTPS:0x007fb272ccf728> (NoMethodError)
        from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/uri/common.rb:623:in `escape'
        from /Users/dmorris/src/mkvm/plugins/autoip.rb:54:in `pre_validate'
        from ./mkvm.rb:88:in `block in <main>'
        from ./mkvm.rb:88:in `each'
        from ./mkvm.rb:88:in `<main>'

File: plugins/autoip.rb [hunk 2]
```diff
@@ -35,29 +36,104 @@ def self.pre_validate options
...
+      uri = URI.escape(uri)
+      uri = URI.parse(uri.concat("addresses/search_hostname/#{options[:hostname]}/"))
+      http = Net::HTTP.new(uri.host, uri.port)
+      http.use_ssl = true
+      request = Net::HTTP::Get.new(uri.request_uri)
+      request.add_field("token",  auth_token)
+      response = http.request(request)
+      if response.code == "404"
+        puts "Requesting IP for #{options[:hostname]}"
+        elsif response.code != "200"
```
Comment (nchowning) [inline plugins/autoip.rb]: Indentation seems a little broken here

Comment (nchowning) [inline plugins/autoip.rb]: I'd also probably change the logic of this section to:
```
if response.code == "404"
  No IP found
elif response.code == "200"
  IP already assigned
else
  Something went wrong
```

File: plugins/autoip.rb [hunk 3]
```diff
@@ -35,29 +36,104 @@ def self.pre_validate options
...
       uri = URI.escape(uri)
-      uri = URI.parse(uri)
+      uri = URI.parse(uri.concat("subnets/cidr/#{options[:subnet]}"))
       http = Net::HTTP.new(uri.host, uri.port)
       http.use_ssl = true
       request = Net::HTTP::Get.new(uri.request_uri)
+      request.add_field("token",  auth_token)
+      response = http.request(request)
+      if response.code =="404"
+          abort "The subnet you requested #{options[:subnet]} can't be found message: #{response.body}"
```
Comment (nchowning) [inline plugins/autoip.rb]: Slightly broken indentation here

**Comment by nchowning** [comment]
I made a few in-line comments about some minor stuff but otherwise lgtm 👍

File: plugins/autoip.rb [hunk 4]
```diff
@@ -35,29 +36,104 @@ def self.pre_validate options
...
       uri = URI.escape(uri)
-      uri = URI.parse(uri)
+      uri = URI.parse(uri.concat("subnets/cidr/#{options[:subnet]}"))
       http = Net::HTTP.new(uri.host, uri.port)
       http.use_ssl = true
       request = Net::HTTP::Get.new(uri.request_uri)
+      request.add_field("token",  auth_token)
+      response = http.request(request)
+      if response.code =="404"
+        abort "The subnet you requested #{options[:subnet]} can't be found message: #{response.body}"
```
Comment (skpy) [inline plugins/autoip.rb]: maybe put a comma after "found" on this output?

File: plugins/autoip.rb [hunk 5]
```diff
@@ -35,29 +36,104 @@ def self.pre_validate options
...
+      uri = URI.escape(uri)
+      uri = URI.parse(uri.concat("subnets/#{subnetId}/first_free/"))
+      http = Net::HTTP.new(uri.host, uri.port)
+      http.use_ssl = true
+      request = Net::HTTP::Get.new(uri.request_uri)
+      request.add_field("token",  auth_token)
       response = http.request(request)
       if response.code != "200"
-        abort "There was an error requesting your IP address, IPAM returned #{response.code}"
+        abort "There was an error while requesting firs free address code: #{response.code}, message: #{response.body}"
```
Comment (skpy) [inline plugins/autoip.rb]: missing a "t" at the end of "first" here.

File: plugins/autoip.rb [hunk 6]
```diff
@@ -35,29 +36,104 @@ def self.pre_validate options
...
+      uri = URI.parse(uri.concat("addresses/?subnetId=#{subnetId}&ip=#{options[:ip]}&hostname=#{options[:hostname]}&owner=#{username}"))
+      http = Net::HTTP.new(uri.host, uri.port)
+      http.use_ssl = true
+      request = Net::HTTP::Post.new(uri.request_uri)
+      request.add_field("token",  auth_token)
+      response = http.request(request)
+      if response.code != "201"
+        abort "There was an error saving the IP and host to the database, returned code: #{response.code}, message: #{response.body}"
       end
-      options[:ip] = response.body
```
Comment (skpy) [inline plugins/autoip.rb]: It doesn't look like you're returning the IP back to the calling module?  You're just outputting it to the terminal?

File: plugins/autoip.rb [hunk 7]
```diff
@@ -35,29 +36,104 @@ def self.pre_validate options
...
+      uri = URI.parse(uri.concat("addresses/search_hostname/#{options[:hostname]}/"))
+      http = Net::HTTP.new(uri.host, uri.port)
+      http.use_ssl = true
+      request = Net::HTTP::Get.new(uri.request_uri)
+      request.add_field("token",  auth_token)
+      response = http.request(request)
+      if response.code == "404"
+        puts "Requesting IP for #{options[:hostname]}"
+      elsif response.code == "200"
+        abort "#{options[:hostname]} is already assigned #{JSON.parse(response.body)["data"][0]["ip"]}"
```
Comment (skpy) [inline plugins/autoip.rb]: By aborting here, you prevent rebuilding a host don't you?  If we delete the VM, but not the host record, we should still be able to rebuild using the original IP, I should think.

File: plugins/autoip.rb (hunk 5 above)
Comment (skpy) [inline plugins/autoip.rb]: And maybe add a comma after "first free address", to make the message a little more legible?

File: plugins/autoip.rb (hunk 6 above)
Comment (dugdigg) [inline plugins/autoip.rb]: When a new IP is needed the assignment happens here:
https://github.com/covermymeds/mkvm/pull/45/files#diff-92f370b6a958a330cb21086a1bcec273R123

I added the assignment needed in existing host search, good catch.

**Comment by dansajner** [comment]
Looks good. 👍   

Please add one more commit to bump the gem version here https://github.com/covermymeds/mkvm/blob/master/lib/mkvm/version.rb

**Pull request merged**